first draft of Pandas Tutor guest post (planned pub date: May 12, 2022) (#27)
pyodide/pyodide-blog: content/posts/pandastutor.md

---
title: "Pandas Tutor: Using Pyodide to Teach Data Science at Scale"

> date: 2022-05-10T08:00:00-07:00

Reviewer: ```suggestion
date: 2022-05-12T08:00:00-07:00
```

for data science, but as teachers we've seen firsthand how it can be
hard for students to learn due to its [complex
API](https://pandas.pydata.org/docs/reference/index.html). For example,
this code selects, sorts, and groups values from a `dogs` dataset to
produce a summary table. But if you run this code in a Jupyter notebook,
all you see is the end result:

<img src="pandastutor-dogs.png" style="width: 70%;"/>

Wait, what exactly did pandas do to turn the `dogs` dataset into this
output summary table? It's not at all clear from what you see above.

If you ran this same code in our [Pandas
Tutor](https://pandastutor.com/) tool, it will show you what's
going on step-by-step as your code selects, sorts, groups, and
calculates the group medians:

<img src="pandastutor-intro-animated.gif"/>

[Try this example live
here](https://pandastutor.com/vis.html#trace=example-code/py_dogs.json).

Instructors can use Pandas Tutor to aid their teaching, and students can
use it to understand and debug their homework assignments.

## Why did we port Pandas Tutor to Pyodide?

The original version of Pandas Tutor ran the user's code on our Linux

> server by spinning up a new Docker container for every code execution request. It main

Reviewer: ```suggestion
server by spinning up a new Docker container for every code execution request. Its main
```

const pandastutor_py = pyodide.pyimport("pandas_tutor.main");
```

This is a synchronous call that imports `pandas_tutor` and all
dependencies when the web app first loads, so it takes 3-5 seconds
to finish since pandas and other libraries must be loaded.
Unfortunately, browser caching can't speed things up here since even if
all files were cached locally, Pyodide still needs to run the import
and initialization code.

One feature addition that might speed this up is if Pyodide
could *serialize a binary memory snapshot* taken right after
everything gets initialized. Then the user's browser can directly load
that snapshot on the first page visit and cache it for future visits.
(But maybe that snapshot would be too big to load efficiently.
[Related GitHub
Issues](https://github.com/pyodide/pyodide/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+size+label%3A%22size+%26+load+time+optimization%22).)

### Step 3: call Pandas Tutor backend code from JavaScript

Now we can call the Pandas Tutor backend from our frontend JavaScript
with just one line:

```js
// 1) frontend grabs userCode from the user's in-browser code editor
// 2) run the user's code to produce an execution trace ...

>       const executionTrace = pandastutor_py.run_user_code(userCode);  

Reviewer: ```suggestion
const executionTrace = pandastutor_py.run_user_code(userCode);  
```
The indentation with respect to the comments looks a bit strange otherwise.

containers. This can be a frustrating user experience when an instructor
is lecturing or when hundreds of students are using the tool at the same
time to visualize in-class code examples.

Beyond user experience, this server-based setup was also a pain for us
as system maintainers since we had to carefully configure our Linux and
Docker setups to handle ever-increasing scale. Remember, we're
university instructors, **not** professional software developers or
DevOps staff! That means we don't have the know-how or institutional
resources required to maintain a production-scale server deployment
setup.

That's why we were thrilled when we discovered
[Pyodide](https://pyodide.org/) since it could help us overcome all
these limitations. Specifically, now that we ported Pandas Tutor over to
use Pyodide:

- Users see their visualizations **instantly** since all their code runs
  in-browser. No more waiting 5-10 seconds for server-side execution and
  network data transfer. A major source of server slowdown was starting
  up a new Python interpreter and loading all dependencies like pandas
  and numpy *every time a user runs a piece of code*. With Pyodide this

>   startup happens only once per browser session, not every execution.

Reviewer: This  point might need to be updated a bit. There is still a startup cost for Pyodide, particularly when loading pandas which I guess should be around 15 s. Then once it's loaded the visualization is indeed fast. 

Also, I understand that the previous setup was much slower, but I'm not sure this could be fully attributed to the interpreter startup,
```
time python -c "import pandas"

real    0m0,220s
user    0m0,345s
sys     0m0,522s
```
on my laptop with SSD, so nowhere near 5s. Also if the code itself runs instantly on pyodide, it would also run instantly on the server since Pyodide is slower than native runtime. So that leaves serializing and network transfer. That adds some overhead, but still 5-10 s feels like a lot.

So my point is running it server-side could theoretically be done with sub second latency given enough engineering resources, but it's also not what you want/can spend your time on (as you say below).

So I think it's good to have this bullet point, but the arguments/wording used might need to be adjusted a bit.

Author: yep, good point, will shorten and adjust. our server setup is fairly sub-optimal due to, as you mentioned, the implementation effort of streamlining it (while keeping things nicely sandboxed, etc.)

`micropip` instead of almost a dozen separate calls:

```js
await pyodide.runPythonAsync(`
import micropip
await micropip.install('https://pandastutor.com/build/pandastutor-1.0-py3-none-any.whl')
`);
```

This approach also has the nice side-effect of pinning all the
dependency versions for better API stability. The only dependency we
didn't bundle was pandas because that requires WASM-compiled code. But
fortunately Pyodide comes with pandas
[built-in](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)!

### Step 2: create a new Pyodide entry point

The only modification we had to make to the Pandas Tutor codebase was
adding a new `pyodide_main.py` file as the entry point for Pyodide:

```py
# simplified contents of pandas_tutor/pyodide_main.py
from .main import run_user_code
import js # to interface with pyodide
def run_code_from_js():
    return run_user_code(js.globalUserCode)
```

This entry point uses the `js` module to access JavaScript global
variables. On the JavaScript end of our web frontend code, we created a
corresponding function that takes the user's code to run (`userCode`)
and calls the `run_code_from_js()` Python function:

```js

> function runPyodidePandastutor(userCode) {
>   window.globalUserCode = userCode; // pass function argument as a global variable
>   const result = pyodide.runPython('pandas_tutor.pyodide_main.run_code_from_js()');
>   window.globalUserCode = undefined; // null this out afterward to prevent leakage
>   return result;
> }

Reviewer: Normally, something like the following would be the way to do it,

```suggestion
function runPyodidePandastutor(userCode) {
  const pandas_tutor_py = pyodide.pyimport("pandas_tutor");
  // assuming main.py is imported in `pandas_tutor.__init__.py` or directly
  // pyimport("pandas_tutor.main")
  const result = pandas_tutor_py.main.run_user_code(user_code);  
  return result;
}
```
and maybe moving the pyimport call outside of the function to avoid running it for each call.

Yeah, ideally it would be nice to make it a bit more elegant since we are showing it in the blog post. If the above doesn't work, let me know and we will find another solution.

```

We couldn't figure out a more elegant way to pass arguments from
JavaScript to Python functions, so we used a `globalUserCode` variable
as a hack that seems to work for now.

### Step 3: import the Pandas Tutor module in JavaScript

Our final setup step was adding a line of code to our frontend
JavaScript to import the Pandas Tutor module:

```js

> pyodide.runPython('import pandas_tutor.pyodide_main');

Reviewer: This would already be done by pyimport.

```

This is a synchronous call that imports `pandas_tutor` and all
dependencies when the web app first loads. This line takes 3-5 seconds
to finish since dependencies (including pandas) must be loaded.
Unfortunately, browser caching can't speed things up here since even if
all the files are cached locally, Pyodide still needs to run the import
and initialization code at page load time.

One feature addition that might speed up this process is if Pyodide
could somehow *serialize a binary memory snapshot* taken right after
everything gets initialized. Then the user's browser can directly load
that snapshot on the first page visit and cache it for future visits.
(But maybe that snapshot would be too big to download efficiently, so

> nevermind!)

Reviewer: The bottleneck for loading pandas is mostly WASM compilation by the browser. It should be cached in Chrome on page reload. But yeah there are certainly lots of ideas to explore in that area https://github.com/pyodide/pyodide/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+size+label%3A%22size+%26+load+time+optimization%22

### Bonus: other fine-tuning and ongoing work

- [Graceful
  degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation):
  To continue supporting older browsers, our code detects if there are
  any problems loading the Pyodide version of Pandas Tutor and, if so,
  automatically reverts back to the original server-side version. Also,
  [loadPyodide](https://pyodide.org/en/stable/usage/api/js-api.html#globalThis.loadPyodide)

>   crashes on some older mobile devices (presumably with [limited

Reviewer: Indeed not enough memory + lack of good WASM support e.g. in Safari on IOS <13

  memory](https://github.com/pyodide/pyodide/issues/533)?) and we can't

>   seem to catch the exception to recover; so we disable it by default on

Reviewer: If you have an example of that, it would be great to open an issue. But yeah, that's unrelated to this post.

  mobile and add a toggle to let the user explicitly enable it. (This is
  also nice since it prevents the page from preemptively downloading
  large amounts of data on users' mobile phone plans.)
- Auto-importer: We want users to be able to import any package when
  writing their code in Pandas Tutor. But Pyodide's
  [loadPackagesFromImports](https://pyodide.org/en/stable/usage/api/js-api.html#pyodide.loadPackagesFromImports)
  works only on packages that come with Pyodide. To give users more
  flexibility, we wrote an auto-importer that automatically calls
  `micropip.install()` whenever the user's code tries to `import` an
  unknown package. These package wheels get downloaded from PyPI
  on-demand and cached in the browser. (Note that this works only if the
  package name on PyPI matches the name in the `import` statement, but

>   this is often true.)

Reviewer: Cool. Just for information that is exactly a reason why we don't auto-install things from PyPi. The other being typos in imports etc. We want Pyodide APIs to be 100% reliable, so that downstream applications can build things on top of them as you are doing here.

Author: yep, totally makes sense!

- Preloading: As soon as the user visits the
  [pandastutor.com](https://pandastutor.com/) landing page, it starts
  downloading Pyodide and selected packages (e.g., pandas, numpy) in the
  background. That way, by the time the user goes to the [visualizer
  page](https://pandastutor.com/vis.html), these
  packages may already have been cached in their browser, so they can
  start working right away. (Thanks to Michael Kennedy from the [Talk
  Python
  podcast](https://talkpython.fm/episodes/show/358/understanding-pandas-visually-with-pandastutor)
  for this suggestion!)
- Caching file downloads: *[work-in-progress]* Some pandas code uses
  functions like
  [read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)
  or
  [read_html](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html)
  to download data files from the web. It would be great to cache those
  files locally in-browser so that when the user re-runs that code in
  the future, Pandas Tutor doesn't need to repeat the same

>   (potentially-large) downloads.

Reviewer: I think if the remote server where the data is downloaded sets the cache headers right, they should already be cached, since we are using fetch requests and these are cached by the browser.

Though yes, maybe manual caching would still be preferable.

- Start-up speed: *[work-in-progress]* Even if all packages are cached,
  it still takes a few seconds to initialize Pandas Tutor at page load
  time. We plan to profile our code and might try using
  [pyc-wheel](https://pypi.org/project/pyc-wheel/) to compile `*.py` in

>   our wheel into `*.pyc` files.

Reviewer: I would be interested in hearing the results.  It's all very related to https://github.com/pyodide/pyodide/issues/347

## The future: Pandas Tutor + Pyodide everywhere!

Pyodide is really exciting for us since it opens the door to **embedding
Pandas Tutor into any existing website** so that anyone can learn data
science within the context of their favorite websites without installing
anything.

To achieve this, we're building a
[bookmarklet](https://en.wikipedia.org/wiki/Bookmarklet) that injects
Pandas Tutor + Pyodide into **any webpage** to add an inline code editor
that can access its DOM. This will allow students to write Python code
to, say, parse HTML tables using
[BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/), clean

> up that raw data, and plot it. Here's a mock-up screenshot of injecting

Reviewer: Very exciting applications!

and plot country populations for a social studies class:

<img src="pandastutor-wiki.png"/>

(This idea originally came from the 2017 research paper, *DS.js: Turn
Any Webpage into an Example-Centric Live Programming Environment for
Learning Data Science* by Xiong Zhang and Philip Guo: [download
PDF](https://pg.ucsd.edu/publications/DSjs-turn-any-webpage-into-data-science-IDE_UIST-2017.pdf))

Next, we can take this idea even further ... wouldn't it be cool if the
official [pandas API
docs](https://pandas.pydata.org/docs/reference/index.html) were
runnable? With this bookmarklet, we can inject Pandas Tutor + Pyodide
into any API docs to automatically parse its code examples, visualize
them, and let students edit that code live to explore different parameter
values:

<img src="pandastutor-api-docs.jpg"/>

We can also do the same with Stack Overflow or other help forum sites
because they often have self-contained code examples that can be
directly visualized:

<img src="pandastutor-stackoverflow.jpg"/>

This is just the beginning of our journey with Pyodide. Stay tuned for
more in the coming months as we use it to build tools to teach data
science at scale!

> -- [Sam](https://www.samlau.me/) and [Philip](https://pg.ucsd.edu/)

Reviewer: You can also add a longer bio as in other posts if you want, but this is also fine. And it's true that you already introduced yourself in the beginning.
